src/D2L.CodeStyle.Analyzers/Language/ImmutableGenericDeclarationAnalyzer.cs

```diff
@@ -67,21 +67,35 @@ KnownImmutableTypes knownImmutableTypes
 					continue;
 				}
 
-				var typeName = typeSymbol.GetFullTypeName();
-				if( typeName == "System.Nullable" ) {
-					var namedTypeSymbol = typeSymbol as INamedTypeSymbol;
-					if (namedTypeSymbol == default(INamedTypeSymbol)) {
-						continue;
-					}
-					typeSymbol = namedTypeSymbol.TypeArguments[ 0 ];
-				}
+				ValidateImmutability(
+					context,
+					typeSymbol,
+					knownImmutableTypes );
+			}
+		}
 
-				if( knownImmutableTypes.IsTypeKnownImmutable( typeSymbol ) ) {
-					continue;
+		private static void ValidateImmutability(
+			SyntaxNodeAnalysisContext context,
+			ITypeSymbol typeSymbol,
+			KnownImmutableTypes knownImmutableTypes
+		) {
+			if( ImmutableContainerMethods.IsAnImmutableContainerType( typeSymbol ) ) {
+
+				var namedTypeSymbol = typeSymbol as INamedTypeSymbol;
+				if( namedTypeSymbol == default( INamedTypeSymbol ) ) {
+					context.ReportDiagnostic( Diagnostic.Create(

```

Comment:
I'm trying to understand this part.  At this point, we are passed a type (which can only be a generic type argument today, but this method shouldn't rely on that).  The type is something like `IEnumerable` that we know to be safe, but we also can't generate an `INamedTypeSymbol`, so we can't analyze type arguments recursively.  Is this possible?  Is it only possible for code that does not compile, yet specifies a fully qualified type name for a System object?

Hard to know whether raising a "Not Immutable" diagnostic makes sense, because I don't know how you could hit this line.

Reply:
It'll force the compile to fail on a type that *should* have been possible to convert to a INamedTypeSymbol, but couldn't for whatever reason.  This is intended to be a fail-safe condition - that is, we will fail the compile rather than letting something go in that we couldn't reason about.

Normally I would throw an exception here, but that would abort all future processing of the analysis, and I still want all the other diagnostics too.

Reply:
Ok, makes sense that if we ever hit this condition, we're forced to stop and look at what went wrong.